File: leetcode/medium/1339-max-prod-splitted/dic.py
"""
  @author Vortexx2
  Problem 1339 - Maximum Product of a Splitted BT

  Runtime - 420 ms
  Memory Usage - 88.2 MB
"""
from typing import Dict, Optional


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def maxProduct(self, root: Optional[TreeNode]) -> int:
    dic = {}
    self.writeSumOfSubtrees(root, dic)
    maxProd = 0

    def helper(fRoot: TreeNode, sRoot: TreeNode):
      if not sRoot:
        return 0
      prod = dic[sRoot] * (dic[fRoot] - dic[sRoot])
      childProd = max(helper(fRoot, sRoot.left), helper(fRoot, sRoot.right))
      return max(prod, childProd)

    maxProd = max(helper(root, root.left), helper(root, root.right))

    return int(maxProd % (1e9 + 7))

  def writeSumOfSubtrees(self, root: TreeNode, dic: Dict) -> None:
    if not root:
      return 0
    dic[root] = self.writeSumOfSubtrees(
        root.left, dic) + self.writeSumOfSubtrees(root.right, dic) + root.val
    return dic[root]

File: leetcode/medium/1339-max-prod-splitted/dic.cpp
/**
 * @file dic.cpp
 * @author Vortexx2
 * @brief Problem 1339 - Max Product of Splitted BT
 * @approach Initial thought using 2 passes
 * @date 21-08-2021
 * 
 * Runtime Error - Because of integer overflow
 */
#include <algorithm>
#include <climits>
#include <deque>
#include <iostream>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  int maxProduct(TreeNode *root) {
    unordered_map<TreeNode *, int> sumDic;
    writesumDic(root, sumDic);
    return (max(helper(root, root->left, sumDic),
                helper(root, root->right, sumDic))) %
           ((long long)1e9 + 7);
  }

  long long helper(TreeNode *fNode, TreeNode *sNode,
                  unordered_map<TreeNode *, int> &sumDic) {
    if (!sNode) return 0;

    long long prod = sumDic[sNode] * (sumDic[fNode] - sumDic[sNode]);
    long long childProd = max(helper(fNode, sNode->left, sumDic),
                             helper(fNode, sNode->right, sumDic));

    return max(prod, childProd);
  }

  int writesumDic(TreeNode *root, unordered_map<TreeNode *, int> &sumDic) {
    if (!root) return 0;

    sumDic.insert({root, writesumDic(root->left, sumDic) +
                             writesumDic(root->right, sumDic) + root->val});

    return sumDic[root];
  }
};